# [실습6] 데이터 정제하기

## 데이터 불러오기

In [60]:
import numpy as np
import pandas as pd

df=pd.read_csv("./data/seoul_park03.csv")
mm=pd.read_csv("./data/misemunji 2.csv")

df.head()

날짜 공휴일     날씨  유료합계    어른    청소년  어린이  외국인   단체  무료합계    총계     연  \
0  2016-01-01   O  구름 조금  3359  2799  141.0  419   47    0  1023  4382  2016   
1  2016-01-02   O  구름 많음  5173  4370  203.0  600  100  111  2092  7265  2016   
2  2016-01-03   O  구름 많음  3008  2571  128.0  309   91    0  1549  4557  2016   
3  2016-01-04   X  구름 많음   890   602    NaN  235   51  223   800  1690  2016   
4  2016-01-05   X  구름 많음   416   319   35.0   62   43   47   840  1256  2016   

   월  일 요일  
0  1  1  금  
1  1  2  토  
2  1  3  일  
3  1  4  월  
4  1  5  화

이번에는 데이터를 다양하게 정제해서 사용하는 방법에 대해 알아보겠습니다.

## 데이터 정렬: sort_values

먼저 데이터를 정렬해서 사용하는 방법에 대해 알아보겠습니다. 현재의 데이터는 시간의 순서대로 기록되어 있는데, 특정 상황에서는 입장객 수에 따라서 정렬을 하면 사용이 용이한 경우가 많습니다. 

이럴 때 우리는 `sort_values()`를 사용할 수 있습니다. `sort_values()`는 특정 컬럼의 값을 기준으로 전체 데이터를 오름차순 혹은 내림차순으로 정렬합니다. 입장객 수가 많은 순서를 알아보기 위해 "총계" 컬럼을 기준으로 내림차순으로 정렬해보도록 하겠습니다.

In [61]:
# 데이터를 확인하기 위해 정렬한 데이터프레임을 df_sorted에 저장
df_sorted=df.sort_values("총계", ascending=False)

# 입장객수가 많았던 날짜의 데이터 상위 10개 출력
df_sorted.head(10)

날짜 공휴일     날씨   유료합계     어른     청소년   어린이  외국인    단체   무료합계  \
125  2016-05-05   O    NaN  40219  38511  1708.0     0  476    35  18469   
755  2018-05-05   O     맑음  29521  28592   899.0    30  208    65  22789   
390  2017-05-05   O  구름 조금  27046  25824  1222.0     0  616   170  19218   
156  2016-06-05   O  구름 조금  33504  28943  1079.0  3482  729   190   8239   
363  2017-04-08   O  구름 조금  28514  24756   638.0  3120  220   137  12373   
378  2017-04-23   O     맑음  25108  21854   269.0  2985  300   333  15008   
364  2017-04-09   O  구름 많음  28496  24523   590.0  3383  283     0  10863   
157  2016-06-06   O    NaN  31030  26358  1423.0  3249  498   392   8241   
413  2017-05-28   O     맑음  28606  25346   675.0  2585  519  4532  10589   
92   2016-04-02   O  구름 조금  31407  26605   561.0  3643  169   201   6380   

        총계     연  월   일 요일  
125  58688  2016  5   5  목  
755  52310  2018  5   5  토  
390  46264  2017  5   5  금  
156  41743  2016  6   5  일  
363  40887  2017  4   8  토  
378  40116  2017  4  23  일  
364  39359  2017  4   9  일  
157  39271  2016  6   6  월  
413  39195  2017  5  28  일  
92   37787  2016  4   2  토

입장객 수가 많았던 순서대로 정렬해보니 데이터가 집계된 기간의 **어린이날**들이 모두 최상위권에 위치해 있어 어린이날의 위엄을 알 수 있습니다. 이는 단순히 어린이날이라는 이유 뿐 만 아니라, 서울대공원의 경우 어린이날 당일에는 13세 미만 어린이들에게 무료 개방을 하고 있기 때문입니다. 이러한 정보를 통해 어린이날의 "어린이" 컬럼의 값이 비정상적으로 낮고 "무료합계" 컬럼이 높은 이유 또한 알 수 있습니다.

또한 전부 공휴일인 것은 모두가 예상하신대로 일 것이고, 어린이날을 제외한 상위권을 살펴보면 **2016년 6월 5일은 다음날이 공휴일인 일요일, 2017년 4월 8일과 9일은 서울대공원 벚꽃축제, 2017년 4월 23일은 서울대공원 겹벚꽃 원더풀 데이 축제 날짜**였습니다. 이렇게 특정 이벤트가 있는 날짜의 입장객 수가 평범한 날들에 비해 월등히 높은 것을 알 수 있습니다.

이번에는 입장객 수 하위 10개 데이터를 살펴보겠습니다.

In [62]:
df_sorted.tail(10)

날짜 공휴일     날씨  유료합계  어른   청소년  어린이  외국인  단체  무료합계   총계     연  월  \
641  2018-01-11   X     맑음    43  34   8.0    1    0   0   269  312  2018  1   
18   2016-01-19   X     맑음    78  61   9.0    8   13   0   222  300  2016  1   
743  2018-04-23   X    눈/비    61  55   2.0    4    0   0   227  288  2018  4   
186  2016-07-05   X    NaN    73  27  45.0    1    4  35   182  255  2016  7   
642  2018-01-12   X     맑음    52  45   5.0    2   15   0   176  228  2018  1   
656  2018-01-26   X     맑음    50  31  13.0    6    0   0   172  222  2018  1   
655  2018-01-25   X     맑음    23  23   0.0    0    1   0   139  162  2018  1   
654  2018-01-24   X     맑음    18  16   1.0    1    0   0   110  128  2018  1   
353  2017-03-29   X  구름 조금     0   0   0.0    0    0   0     7    7  2017  3   
352  2017-03-28   X  구름 많음     0   0   0.0    0    0   0     2    2  2017  3   

      일 요일  
641  11  목  
18   19  화  
743  23  월  
186   5  화  
642  12  금  
656  26  금  
655  25  목  
654  24  수  
353  29  수  
352  28  화

하위 2개 데이터는 폐쇄 후 재개장을 위한 테스트로 추정되기 때문에 제외하고, 나머지 데이터를 살펴보면 일단 전부 공휴일이 아니라는 것을 알 수 있습니다. 최하위권에 **2018년 1월 24일부터 25일까지**가 위치해있는데, 날씨가 맑음임에도 불구하고 아무리 평일이라도 입장객 수가 지나치게 낮은 것을 확인할 수 있습니다. 이유를 알아보면 해당 일차의 날씨 자체는 맑음이 맞지만 **강렬한 한파로 인해 서울의 최저기온이 영하 18도까지 내려갔던 기간**이었습니다.

이렇게 데이터를 정렬하면 추가 정보를 얻을 수 있고, 데이터 분석의 방향을 결정할 수도 있습니다.

### [TODO] 미세먼지 데이터에서 초미세먼지 농도가 3번째로 심했던 날짜를 변수 `pollution`에 저장하세요.
* `mm`에는 미세먼지 데이터가 데이터프레임 형태로 저장되어있습니다.
* `sort_values`를 활용해 "초미세먼지" 컬럼을 기준으로 내림차순 정렬하고, 3번째 데이터를 확인합니다.
* 초미세먼지 농도가 3번째로 높았던 날짜를 `"yyyymmdd"` 형식의 문자열로 변수 `pollution`에 저장하세요. (ex. 2016년 1월 5일 -> `pollution="20160105"`)

In [63]:
# sort_values를 활용해 데이터프레임 mm을 "초미세먼지" 컬럼 기준으로 내림차순 정렬하세요.
mm.sort_values('초미세먼지',ascending=False)

날짜  미세먼지  초미세먼지     오존  이산화질소  일산화탄소  아황산가스
1159  2019-03-05   186    135  0.035  0.036    0.9  0.006
1109  2019-01-14   166    129  0.005  0.064    1.4  0.006
1158  2019-03-04   169    117  0.030  0.050    1.0  0.006
1160  2019-03-06   141    101  0.028  0.044    0.9  0.006
814   2018-03-25   121     99  0.038  0.035    0.8  0.005
...          ...   ...    ...    ...    ...    ...    ...
999   2018-09-26     8      4  0.029  0.010    0.3  0.003
1000  2018-09-27     9      4  0.026  0.015    0.3  0.003
593   2017-08-16     7      4  0.012  0.014    0.3  0.004
917   2018-07-06     7      3  0.020  0.010    0.3  0.003
1009  2018-10-06     6      3  0.027  0.014    0.3  0.003

[1186 rows x 7 columns]

In [74]:
#loc 쓰면 기존 인덱스 유지되서 reset 시키거나 iloc 써야함
mm.sort_values('초미세먼지', ascending=False).iloc[2]['날짜'].replace("-","")

'20190304'

In [75]:
# 초미세먼지 농도가 3번째로 높은 날의 날짜정보를 변수 pollution에 저장하세요.
# None을 지우고 알맞은 값을 입력하세요.
pollution =mm.sort_values('초미세먼지', ascending=False).iloc[2]['날짜'].replace("-","")
(pollution)

'20190304'

## 인덱스 재지정: reset_index

이런식으로 데이터를 정렬하여 데이터를 사용하다보면 한 가지 문제가 발생합니다. 다시 `df_sorted`를 살펴보시면 데이터는 "총계"컬럼에 맞춰서 정렬되었지만 이 과정에서 기존의 인덱스를 그대로 가지고 정렬된 사실을 확인할 수 있습니다. 이렇게 될 경우 앞서 배웠던 `loc` 등을 활용하기가 어렵기 때문에, 이 데이터를 가지고 계속 데이터 분석을 할 예정이라면 인덱스를 재정렬해줄 필요가 있습니다. 이 때 우리는 `reset_index()`를 활용합니다.

In [76]:
df_sorted.reset_index(drop=True) # drop을 True로 지정해주어 "index"컬럼 생성 안함

날짜 공휴일     날씨   유료합계     어른     청소년   어린이  외국인   단체   무료합계  \
0     2016-05-05   O    NaN  40219  38511  1708.0     0  476   35  18469   
1     2018-05-05   O     맑음  29521  28592   899.0    30  208   65  22789   
2     2017-05-05   O  구름 조금  27046  25824  1222.0     0  616  170  19218   
3     2016-06-05   O  구름 조금  33504  28943  1079.0  3482  729  190   8239   
4     2017-04-08   O  구름 조금  28514  24756   638.0  3120  220  137  12373   
...          ...  ..    ...    ...    ...     ...   ...  ...  ...    ...   
1081  2018-01-26   X     맑음     50     31    13.0     6    0    0    172   
1082  2018-01-25   X     맑음     23     23     0.0     0    1    0    139   
1083  2018-01-24   X     맑음     18     16     1.0     1    0    0    110   
1084  2017-03-29   X  구름 조금      0      0     0.0     0    0    0      7   
1085  2017-03-28   X  구름 많음      0      0     0.0     0    0    0      2   

         총계     연  월   일 요일  
0     58688  2016  5   5  목  
1     52310  2018  5   5  토  
2     46264  2017  5   5  금  
3     41743  2016  6   5  일  
4     40887  2017  4   8  토  
...     ...   ... ..  .. ..  
1081    222  2018  1  26  금  
1082    162  2018  1  25  목  
1083    128  2018  1  24  수  
1084      7  2017  3  29  수  
1085      2  2017  3  28  화  

[1086 rows x 15 columns]

`df_sorted`의 인덱스가 깔끔하게 재정렬된 것을 확인할 수 있습니다. 이렇게 인덱스를 재지정은 지금처럼 데이터를 정렬하거나, 일부 데이터를 삭제하는 등의 작업을 통해 인덱스가 흐트러졌을 때 수행해주면 유용한 작업입니다.

## 데이터 삭제: drop

데이터에는 다양한 정보가 저장되어 있지만, 데이터분석 목적에 따라 크게 중요하지 않은 정보들이 있을 수 있습니다. 가령 서울대공원 데이터를 다시 살펴보면 "유료합계" 컬럼과 "무료합계" 컬럼은 입장객 수를 분석하는 데에 있어서 크게 중요하지 않습니다. 이런 경우 해당 컬럼을 `drop()`을 활용하여 삭제하여 좀 더 깔끔하고 명확한 데이터를 분석하는데 활용할 수 있습니다. "유료합계"와 "무료합계" 컬럼을 삭제하여 `df`에 덮어씌우는 과정은 다음과 같습니다.

In [77]:
# 삭제할 2개의 컬럼을 리스트로 입력
# axis=1로 설정하여 컬럼을 삭제
df=df.drop(["유료합계", "무료합계"], axis=1)
df

날짜 공휴일     날씨    어른    청소년  어린이  외국인   단체    총계     연  월   일 요일
0     2016-01-01   O  구름 조금  2799  141.0  419   47    0  4382  2016  1   1  금
1     2016-01-02   O  구름 많음  4370  203.0  600  100  111  7265  2016  1   2  토
2     2016-01-03   O  구름 많음  2571  128.0  309   91    0  4557  2016  1   3  일
3     2016-01-04   X  구름 많음   602    NaN  235   51  223  1690  2016  1   4  월
4     2016-01-05   X  구름 많음   319   35.0   62   43   47  1256  2016  1   5  화
...          ...  ..    ...   ...    ...  ...  ...  ...   ...   ... ..  .. ..
1081  2019-03-27   X  구름 많음   464   10.0   30   21    0  1117  2019  3  27  수
1082  2019-03-28   X  구름 많음   687   46.0   28   35  108  1665  2019  3  28  목
1083  2019-03-29   X  구름 조금  1447  120.0   77   14  188  2870  2019  3  29  금
1084  2019-03-30   O     흐림  1326   44.0  169   29  115  2452  2019  3  30  토
1085  2019-03-31   O  구름 조금  2563  111.0  387   53    0  4418  2019  3  31  일

[1086 rows x 13 columns]

필요 없는 컬럼이 삭제되어 좀더 깔끔한 데이터가 된 것을 확인할 수 있습니다. 

## 열 이름 바꾸기: rename

떄로는 데이터의 컬럼이름이 잘못 설정되어있거나, 알아보기 어렵게 설정되어있는 경우가 있습니다. 이런 경우 `rename()`을 활용해 컬럼의 이름을 바꿀 수 있습니다. 서울대공원 데이터에서 "총계" 컬럼의 경우 총 관람객 수를 의미한다는 것을 유추할 수 있지만, 좀 더 명확하게 "총계" 컬럼의 이름을 "총입장객수" 로 바꿔보도록 하겠습니다.

In [78]:
df=df.rename(columns = {"총계": "총입장객수"})
df

날짜 공휴일     날씨    어른    청소년  어린이  외국인   단체  총입장객수     연  월   일 요일
0     2016-01-01   O  구름 조금  2799  141.0  419   47    0   4382  2016  1   1  금
1     2016-01-02   O  구름 많음  4370  203.0  600  100  111   7265  2016  1   2  토
2     2016-01-03   O  구름 많음  2571  128.0  309   91    0   4557  2016  1   3  일
3     2016-01-04   X  구름 많음   602    NaN  235   51  223   1690  2016  1   4  월
4     2016-01-05   X  구름 많음   319   35.0   62   43   47   1256  2016  1   5  화
...          ...  ..    ...   ...    ...  ...  ...  ...    ...   ... ..  .. ..
1081  2019-03-27   X  구름 많음   464   10.0   30   21    0   1117  2019  3  27  수
1082  2019-03-28   X  구름 많음   687   46.0   28   35  108   1665  2019  3  28  목
1083  2019-03-29   X  구름 조금  1447  120.0   77   14  188   2870  2019  3  29  금
1084  2019-03-30   O     흐림  1326   44.0  169   29  115   2452  2019  3  30  토
1085  2019-03-31   O  구름 조금  2563  111.0  387   53    0   4418  2019  3  31  일

[1086 rows x 13 columns]

컬럼의 이름이 "총입장객수" 로 바뀐것을 확인할 수 있습니다. 

## 결측치 처리

우리가 실습에서 사용하고 있는 데이터는 실제 서울대공원의 입장객 수를 기록한 데이터입니다. 그런데 실제로 데이터를 수집하다보면 다양한 문제가 야기됩니다. 앞서 처리했던 표기의 문제도 있었고, 데이터 자료형의 문제도 있었습니다. 그리고 또 다른 문제로는 데이터의 누락이 있습니다. 오랜 기간 데이터를 수집하다보면 데이터가 수집되지 않는 경우들이 발생할 수 있습니다.


`info()`를 활용하여 다시 한번 데이터를 살펴보겠습니다.

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1086 entries, 0 to 1085
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   날짜      1086 non-null   object 
 1   공휴일     1086 non-null   object 
 2   날씨      946 non-null    object 
 3   어른      1086 non-null   int64  
 4   청소년     1081 non-null   float64
 5   어린이     1086 non-null   int64  
 6   외국인     1086 non-null   int64  
 7   단체      1086 non-null   int64  
 8   총입장객수   1086 non-null   int64  
 9   연       1086 non-null   int64  
 10  월       1086 non-null   int64  
 11  일       1086 non-null   int64  
 12  요일      1086 non-null   object 
dtypes: float64(1), int64(8), object(4)
memory usage: 110.4+ KB


데이터의 갯수를 보면 대부분의 컬럼이 1086개의 데이터를 가지고 있지만 "날씨" 컬럼과 "청소년" 컬럼의 데이터가 부족한것을 확인할 수 있습니다. 잘보면 데이터의 갯수가 Non-Null Count라고 되어있는데, 즉 "날씨" 컬럼과 "청소년" 컬럼에는 Null이 존재한다는 뜻입니다.

이 Null이 어떻게 생겼는지는 데이터프레임에서 확인할 수 있습니다.

In [80]:
df.head()

날짜 공휴일     날씨    어른    청소년  어린이  외국인   단체  총입장객수     연  월  일 요일
0  2016-01-01   O  구름 조금  2799  141.0  419   47    0   4382  2016  1  1  금
1  2016-01-02   O  구름 많음  4370  203.0  600  100  111   7265  2016  1  2  토
2  2016-01-03   O  구름 많음  2571  128.0  309   91    0   4557  2016  1  3  일
3  2016-01-04   X  구름 많음   602    NaN  235   51  223   1690  2016  1  4  월
4  2016-01-05   X  구름 많음   319   35.0   62   43   47   1256  2016  1  5  화

2016년 1월 4일의 데이터의 "청소년" 컬럼을 확인하면 NaN(Not a Number)라고 되어있습니다. 이는 0이 아니라 아예 공백, 즉 수집되지 않아서 어떤 값인지 알 수 없는 데이터입니다. 이런 값들을 **결측치** 라고 합니다.

당연한 얘기지만, 이런 결측치들은 데이터 분석이나 인공지능 적용 과정에서 걸림돌로 작용합니다. 이제부터 이 결측치를 처리하는 방법을 배워보겠습니다.

## 결측치 탐색: isnull()

먼저 결측치가 얼마나 존재하는지 알아보기 위해 `isnull()`을 활용합니다. 물론 앞서 봤듯이 `info()`를 활용해서도 알 수 있지만, `isnull()`을 활용하면 추가적인 계산 필요 없이 직관적으로 어떤 컬럼에 몇 개의 결측치가 존재하는지 확인할 수 있습니다. 먼저 데이터프레임에 `isnull()`을 적용해보겠습니다.

In [81]:
df.isnull()

날짜    공휴일     날씨     어른    청소년    어린이    외국인     단체  총입장객수      연  \
0     False  False  False  False  False  False  False  False  False  False   
1     False  False  False  False  False  False  False  False  False  False   
2     False  False  False  False  False  False  False  False  False  False   
3     False  False  False  False   True  False  False  False  False  False   
4     False  False  False  False  False  False  False  False  False  False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
1081  False  False  False  False  False  False  False  False  False  False   
1082  False  False  False  False  False  False  False  False  False  False   
1083  False  False  False  False  False  False  False  False  False  False   
1084  False  False  False  False  False  False  False  False  False  False   
1085  False  False  False  False  False  False  False  False  False  False   

          월      일     요일  
0     False  False  False  
1     False  False  False  
2     False  False  False  
3     False  False  False  
4     False  False  False  
...     ...    ...    ...  
1081  False  False  False  
1082  False  False  False  
1083  False  False  False  
1084  False  False  False  
1085  False  False  False  

[1086 rows x 13 columns]

`isnull()`은 데이터프레임의 각 원소가 결측치인지 아닌지를 검사합니다. 결측치가 아닌 제대로된 값은 False, 결측치는 True로 채워진것을 확인할 수 있습니다. 하지만 저 표를 가지고는 결측치가 컬럼별로 몇개인지 확인할 수가 없습니다. 이 때 True는 1로, False는 0으로 계산되어 `sum()`을 활용해서 각 컬럼별로 전체값의 합을 구하면 결과값은 컬럼에 존재하는 True의 갯수, 즉 결측치의 갯수가 됩니다.

In [82]:
df.isnull().sum()

날짜         0
공휴일        0
날씨       140
어른         0
청소년        5
어린이        0
외국인        0
단체         0
총입장객수      0
연          0
월          0
일          0
요일         0
dtype: int64

"날씨" 컬럼에 140개, "청소년" 컬럼에 5개의 결측치가 있는 것을 알 수 있습니다. 이렇게 결측치의 갯수를 파악하는 것은 결측치를 어떻게 처리할지 방법을 결정하는데에 중요합니다.

결측치를 처리하는 방법은 대표적으로 2가지가 있습니다. 하나는 결측치를 **특정 값으로 채워넣는 것**이고, 하나는 **결측치가 존재하는 데이터를 삭제하는 것**입니다. 

## 결측치 채우기: fillna()

먼저 결측치를 채워넣는 방법에 대해 알아보겠습니다. `fillna()`는 결측치를 특정 값으로 채워넣는데 활용하는 메서드로, 어떤 값으로 채워넣는지는 다양한 방법을 사용합니다. 이번 실습에서는 "청소년" 컬럼의 결측치를 전체 데이터의 청소년 입장객의 평균값으로 채워보도록 하겠습니다.

In [83]:
df.fillna(int(df["청소년"].mean()))

날짜 공휴일     날씨    어른    청소년  어린이  외국인   단체  총입장객수     연  월   일 요일
0     2016-01-01   O  구름 조금  2799  141.0  419   47    0   4382  2016  1   1  금
1     2016-01-02   O  구름 많음  4370  203.0  600  100  111   7265  2016  1   2  토
2     2016-01-03   O  구름 많음  2571  128.0  309   91    0   4557  2016  1   3  일
3     2016-01-04   X  구름 많음   602  463.0  235   51  223   1690  2016  1   4  월
4     2016-01-05   X  구름 많음   319   35.0   62   43   47   1256  2016  1   5  화
...          ...  ..    ...   ...    ...  ...  ...  ...    ...   ... ..  .. ..
1081  2019-03-27   X  구름 많음   464   10.0   30   21    0   1117  2019  3  27  수
1082  2019-03-28   X  구름 많음   687   46.0   28   35  108   1665  2019  3  28  목
1083  2019-03-29   X  구름 조금  1447  120.0   77   14  188   2870  2019  3  29  금
1084  2019-03-30   O     흐림  1326   44.0  169   29  115   2452  2019  3  30  토
1085  2019-03-31   O  구름 조금  2563  111.0  387   53    0   4418  2019  3  31  일

[1086 rows x 13 columns]

원래는 결측치였던 2016년 1월 4일의 청소년 데이터가 463명이라는 펑균값으로 채워진것을 확인할 수 있습니다. 물론 이 값이 정확한 값이라고는 할 수 없지만, 최대한 확률이 높은 값으로 채워넣었다고 생각하면 됩니다.

## 결측치 데이터 삭제하기: dropna()

결측치가 많지 않을 경우, 결측치가 있는 데이터를 삭제해버리는 것도 좋은 방법입니다. 가령 청소년 컬럼에서 결측치가 있는 데이터를 삭제하면 5일치의 데이터가 사라지는 것입니다. 이는 전체 기간 1086일의 데이터에서 비중이 크지 않기 때문에 괜찮은 방법이 될 수도 있습니다. 이럴 때 `dropna()`를 활용하면 알아서 기준 컬럼에서 결측치를 탐사하고 해당하는 데이터를 삭제합니다.

In [84]:
df.dropna(subset="청소년")

날짜 공휴일     날씨    어른    청소년  어린이  외국인   단체  총입장객수     연  월   일 요일
0     2016-01-01   O  구름 조금  2799  141.0  419   47    0   4382  2016  1   1  금
1     2016-01-02   O  구름 많음  4370  203.0  600  100  111   7265  2016  1   2  토
2     2016-01-03   O  구름 많음  2571  128.0  309   91    0   4557  2016  1   3  일
4     2016-01-05   X  구름 많음   319   35.0   62   43   47   1256  2016  1   5  화
5     2016-01-06   X  구름 조금   277   41.0   49   34    0   1174  2016  1   6  수
...          ...  ..    ...   ...    ...  ...  ...  ...    ...   ... ..  .. ..
1081  2019-03-27   X  구름 많음   464   10.0   30   21    0   1117  2019  3  27  수
1082  2019-03-28   X  구름 많음   687   46.0   28   35  108   1665  2019  3  28  목
1083  2019-03-29   X  구름 조금  1447  120.0   77   14  188   2870  2019  3  29  금
1084  2019-03-30   O     흐림  1326   44.0  169   29  115   2452  2019  3  30  토
1085  2019-03-31   O  구름 조금  2563  111.0  387   53    0   4418  2019  3  31  일

[1081 rows x 13 columns]

이번에는 "청소년" 컬럼에 결측치가 있었던 2016년 1월 4일의 데이터가 아예 삭제된 것을 확인할 수 있습니다. 

이렇게 결측치를 처리하는 방법은 다양하지만, 이번 강의에서는 이후 강의진행의 편의성을 위해 결측치를 채워넣는 방식으로 `df`를 변형하도록 하겠습니다.

In [85]:
df=df.fillna(int(df["청소년"].mean()))

### [TODO] 미세먼지 데이터에서 결측치가 있는 컬럼을 파악하고 해당 컬럼의 중앙값으로 채운 뒤 `mm`에 덮어씌워 저장하세요.
* 미세먼지 데이터가 담겨있는 데이터프레임 `mm`의 한 컬럼에 결측치가 4개 존재합니다. 
* 먼저 `isnull()`과 `sum()`을 활용하여 어떤 컬럼에 결측치가 존재하는지 확인하세요.
* 결측치를 **결측치가 있는 컬럼 전체의 중앙값**으로 채운 뒤, `mm`의 해당 컬럼에 덮어씌워 저장하세요. 중간값 집계함수는 `median()`입니다.
* 모든 지시사항을 수행한 뒤 데이터프레임 `mm`에는 결측치가 없어야 합니다.

In [88]:
# 미세먼지 데이터를 다시 불러와 변수 mm에 저장합니다.
mm=pd.read_csv("./data/misemunji 2.csv")
mm.head()

날짜  미세먼지  초미세먼지     오존  이산화질소  일산화탄소  아황산가스
0  2016-01-01    68     45  0.006  0.047    1.0  0.006
1  2016-01-02    51     31  0.009  0.044    0.7  0.006
2  2016-01-03    96     58  0.013  0.049    0.9  0.007
3  2016-01-04    74     47  0.025  0.025    0.6  0.006
4  2016-01-05    32     16  0.016  0.025    0.4  0.006

In [89]:
# isnull()과 sum()을 활용해 어떤 컬럼에 결측치가 존재하는지 확인하세요.

mm.isnull().sum()

날짜       0
미세먼지     0
초미세먼지    0
오존       0
이산화질소    4
일산화탄소    0
아황산가스    0
dtype: int64

In [91]:
# mm의 결측치를 지시사항에 맞게 채워넣은 뒤 덮어씌우세요.
mm = mm.fillna(mm['이산화질소'].median())

날짜       0
미세먼지     0
초미세먼지    0
오존       0
이산화질소    0
일산화탄소    0
아황산가스    0
dtype: int64